# «Секреты Темнолесья»: исследование игровой индустрии 2000–2013

### Описание проекта
Проект команды «Секреты Темнолесья» — это исследование развития видеоигр с 2000 по 2013 год, нацеленное на привлечение новой аудитории, неравнодушной к классическим играм. В рамках проекта анализируются исторические данные о продажах игр в разных жанрах и на разных платформах, региональные предпочтения игроков, а также пользовательские и экспертные оценки. Особый акцент сделан на жанре RPG — компьютерных ролевых играх, где игроки управляют персонажами в проработанных мирах.

Результатом проекта станет аналитическая статья-исследование, которая будет опубликована на известном IT-ресурсе. Статья охватывает ключевые платформы эпохи (PlayStation 2, Xbox, Nintendo Wii и другие), эволюцию игровых жанров, рост многопользовательских онлайн-сообществ и превращение видеоигр в важную часть массовой культуры. Цель — не только рассказать об истории индустрии, но и познакомить читателей с миром «Секретов Темнолесья».

### Цели и задачи проекта
В этой тетрадке используем датасет `new_games.csv`, который информацию о продажах игр разных жанров и платформ, а также пользовательские и экспертные оценки игр.
Основная задача - познакомиться с данными, проверить их корректность и провести предобработку, получив необходимый срез данных.
Доплючательные задачи:
- Отобрать данные по времени выхода игры. Вам нужен период с 2000 по 2013 год включительно.

- Категоризовать игры по оценкам пользователей и экспертов. Выделите три категории:

    - высокая оценка — с оценкой от 8 до 10 и от 80 до 100, включая правые границы интервалов.

    - средняя оценка — с оценкой от 3 до 8 и от 30 до 80, не включая правые границы интервалов.
    
    - низкая оценка — с оценкой от 0 до 3 и от 0 до 30, не включая правые границы интервалов.

- Выделить топ-7 платформ по количеству игр, выпущенных за весь требуемый период.

### Описание данных
Данные `new_games.csv` содержат информацию о продажах игр разных жанров и платформ, а также пользовательские и экспертные оценки игр:
- `Name` — название игры.
- `Platform` — название платформы.
- `Year of Release` — год выпуска игры.
- `Genre` — жанр игры.
- `NA sales` — продажи в Северной Америке (в миллионах проданных копий).
- `EU sales` — продажи в Европе (в миллионах проданных копий).
- `JP sales` — продажи в Японии (в миллионах проданных копий).
- `Other sales` — продажи в других странах (в миллионах проданных копий).
- `Critic Score` — оценка критиков (от 0 до 100).
- `User Score` — оценка пользователей (от 0 до 10).
- `Rating` — рейтинг организации ESRB (англ. Entertainment Software Rating Board). Эта ассоциация определяет рейтинг компьютерных игр и присваивает им подходящую возрастную категорию.


## Загрузка и знакомство с данными:

Загрузим необходимые библиотеки для анализа данных и данные из датасета `new_games.csv`. Затем выведем основную информацию о данных с помощью метода `info()` и первые строки датафрейма.

In [1]:
# Импортируем библиотеку pandas
import pandas as pd

In [2]:
# Выгружаем данные из датасета new_games.csv в датафрейм games
games = pd.read_csv('https://code.s3.yandex.net/datasets/new_games.csv')

In [3]:
# Выводим информацию о датафрейме
games.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


In [4]:
# Выводим первые 5 строк датафрейма на экран
games.head()

,Name,Platform,Year of Release,Genre,NA sales,EU sales,JP sales,Other sales,Critic Score,User Score,Rating
0,Wii Sports,Wii,2006.0,Sports,41.36,28.96,3.77,8.45,76.0,8,E
1,Super Mario Bros.,NES,1985.0,Platform,29.08,3.58,6.81,0.77,NaN,NaN,NaN
2,Mario Kart Wii,Wii,2008.0,Racing,15.68,12.76,3.79,3.29,82.0,8.3,E
3,Wii Sports Resort,Wii,2009.0,Sports,15.61,10.93,3.28,2.95,80.0,8,E
4,Pokemon Red/Pokemon Blue,GB,1996.0,Role-Playing,11.27,8.89,10.22,1.00,NaN,NaN,NaN


In [5]:
# Выводим последние 5 строк датафрейма
games.tail()

,Name,Platform,Year of Release,Genre,NA sales,EU sales,JP sales,Other sales,Critic Score,User Score,Rating
16951,Samurai Warriors: Sanada Maru,PS3,2016.0,Action,0.00,0.0,0.01,0.0,NaN,NaN,NaN
16952,LMA Manager 2007,X360,2006.0,Sports,0.00,0.01,0.0,0.0,NaN,NaN,NaN
16953,Haitaka no Psychedelica,PSV,2016.0,Adventure,0.00,0.0,0.01,0.0,NaN,NaN,NaN
16954,Spirits & Spells,GBA,2003.0,Platform,0.01,0.0,0.0,0.0,NaN,NaN,NaN
16955,Winning Post 8 2016,PSV,2016.0,Simulation,0.00,0.0,0.01,0.0,NaN,NaN,NaN


In [6]:
# Выводим случайные 5 строк датафрейма
games.sample(5)

,Name,Platform,Year of Release,Genre,NA sales,EU sales,JP sales,Other sales,Critic Score,User Score,Rating
12155,Machine Head,PS,1996.0,Shooter,0.04,0.03,0.0,0.00,NaN,NaN,NaN
12794,Resident Evil - Code: Veronica X,GC,2003.0,Action,0.05,0.01,0.0,0.00,62.0,8.5,M
16040,Aedis Eclipse: Generation of Chaos,PSP,2006.0,Role-Playing,0.02,0.0,0.0,0.00,62.0,6.5,E10+
16371,Total War Attila: Tyrants & Kings,PC,2016.0,Strategy,0.00,0.01,0.0,0.00,NaN,NaN,NaN
5777,Batman: Arkham Origins,PC,2013.0,Action,0.13,0.16,0.0,0.02,74.0,7.5,T


Датасет `new_games.csv` содержит 11 столбцов и 16956 строк, в которых представлена информация о продажах игр разных жанров и платформ, а также пользовательские и экспертные оценки игр.

Изучим типы данных и их корректность:
- **Числовые значения с плавающей запятой (float64).** Четыре столбца имеют тип данных `float64`:

    - `Year of Release`, `NA sales`, `Other sales` и `Critic Score` содержат числовые значения с плавающей запятой.
    
    Столбец `Year of Release` содержит год выхода игры, а столбцы `NA sales`, `Other sales` и `Critic Score` содержат количество продаж игры в Северной Америке, в иных регионах и оценках критиках. Столбец `Year of Release` стоит привести к типу `int16`, а столбцы `NA sales`, `Other sales` и `Critic Score` - к типу `float32`, чтобы уменьшить размер данных и улучшить производительность. 

- **Строковые данные (object).** Семь столбцов имеют тип данных `object`:
    - `Name`, `Platform`, `Genre`, `Rating` содержат строковую информацию (название игры, игровую платформу, жанр и рейтинг в буквенном выражении). Здесь тип данных `object` корректный.
    - `EU sales`, `JP sales`, `User Score` хранят информацию о продажах игры в Европе и Японии, а также оценки пользователей. Здесь тип данных `object` не корректный. Для таких данных рекомендуется использовать тип `float32`.

Названия столбцов приведем к "зменным именам", чтобы было удобнее их использовать в дальнейшем.

В данных встречаются пропуски. Наибольшее их количество в столбцах оценок и рейтинге.

## Проверка ошибок в данных и их предобработка

### Названия столбцов датафрейма

In [7]:
# Выводим название столбцов
list(games.columns)

['Name',
 'Platform',
 'Year of Release',
 'Genre',
 'NA sales',
 'EU sales',
 'JP sales',
 'Other sales',
 'Critic Score',
 'User Score',
 'Rating']

Названия столбцов приведем к "зменным именам", чтобы было удобнее их использовать в дальнейшем.

In [8]:
# Приводим названия столбцов к "змеиным именам"
games.columns = games.columns.str.lower().str.replace(" ", "_")

In [9]:
# Выводим список столбцов после изменения
list(games.columns)

['name',
 'platform',
 'year_of_release',
 'genre',
 'na_sales',
 'eu_sales',
 'jp_sales',
 'other_sales',
 'critic_score',
 'user_score',
 'rating']

In [10]:
# Выводим первые 5 строк датафрейма
games.head()

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating
0,Wii Sports,Wii,2006.0,Sports,41.36,28.96,3.77,8.45,76.0,8,E
1,Super Mario Bros.,NES,1985.0,Platform,29.08,3.58,6.81,0.77,NaN,NaN,NaN
2,Mario Kart Wii,Wii,2008.0,Racing,15.68,12.76,3.79,3.29,82.0,8.3,E
3,Wii Sports Resort,Wii,2009.0,Sports,15.61,10.93,3.28,2.95,80.0,8,E
4,Pokemon Red/Pokemon Blue,GB,1996.0,Role-Playing,11.27,8.89,10.22,1.00,NaN,NaN,NaN


### Наличие пропусков в данных

В этом разделе посмотрим на количество пропущенных значений по столбцам датафрейма.

In [11]:
# Выводим количество пропущенных строк в датафрейме
games.isna().sum()

name                  2
platform              0
year_of_release     275
genre                 2
na_sales              0
eu_sales              0
jp_sales              0
other_sales           0
critic_score       8714
user_score         6804
rating             6871
dtype: int64

In [12]:
# Подсчитываем процент строк с пропусками
games.isna().sum() / len(games) * 100

name                0.011795
platform            0.000000
year_of_release     1.621845
genre               0.011795
na_sales            0.000000
eu_sales            0.000000
jp_sales            0.000000
other_sales         0.000000
critic_score       51.391838
user_score         40.127389
rating             40.522529
dtype: float64

## Промежуточный вывод по пропущенным значениям

В датафрейме `games` пропуски присутствуют в следующих столбцах:

| Столбец | Количество пропусков | Доля от всех строк |
|---------|----------------------|---------------------|
| `name` | 2 | 0.0118% |
| `year_of_release` | 275 | 1.6218% |
| `genre` | 2 | 0.0118% |
| `critic_score` | 8714 | 51.3918% |
| `user_score` | 6804 | 40.1274% |
| `rating` | 6871 | 40.5225% |

Столбцы `platform`, `na_sales`, `eu_sales`, `jp_sales`, `other_sales` пропусков не содержат.

- **`name`, `genre`** — единичные пропуски (по 2 строки). Скорее всего, это технические ошибки при сборе или выгрузке данных, либо потеря информации по конкретным играм.
- **`year_of_release`** — пропущено 275 значений (1.62%). Вероятные причины: для некоторых игр год выпуска неизвестен (например, очень старые или малоизвестные проекты), либо данные не были указаны в источнике.
- **`critic_score`, `user_score`, `rating`** — пропуски очень массовые (40–51%). Причины: игры могут быть выпущены давно, не иметь достаточного количества отзывов, быть региональными или нишевыми. Столбец `rating` часто является производным от оценок (например, ESRB), поэтому если нет оценок, то и рейтинг может отсутствовать.

#### Действия с пропущенными данными

- **Для `name` и `genre`** (по 2 пропуска):  
  Можно удалить эти строки, так как потеря составит ничтожную долю (0.01%) и не повлияет на общую статистику.

- **Для `year_of_release`** (275 пропусков, 1.62%):  
  По категории платформы можно заполнить медианным годом.

- **Для `critic_score`, `user_score`, `rating`** (40–51% пропусков):  
  Удаление строк приведёт к потере почти половины датасета, что недопустимо. Можно заменить на значение '-1' с целью использовать его как отсутствие информации.

### Обработка пропущенных значений и аномалий

In [13]:
# Удаляем строки без названия игры (всего 2)
games = games.dropna(subset=['name'])

In [14]:
# year_of_release можно заполнить медианой по платформе
games['year_of_release'] = games['year_of_release'].fillna(
games.groupby('platform')['year_of_release'].transform('median'))

In [15]:
# Заполняем рейтинги значением '-1' / -1
games['critic_score'] = games['critic_score'].fillna(-1)
games[['rating', 'user_score']] = (
    games[['rating', 'user_score']].fillna('-1')
)

In [16]:
# Обработка аномальных значений в столбцах с выручкой
games[['eu_sales', 'jp_sales', 'other_sales']] = games[['eu_sales', 'jp_sales', 'other_sales']].replace(['', 'NA', 'N/A', 'null', '?', '-', 'unknown'], '')

In [17]:
# Проверяем, что games очистился
games.isna().sum()

name               0
platform           0
year_of_release    0
genre              0
na_sales           0
eu_sales           0
jp_sales           0
other_sales        0
critic_score       0
user_score         0
rating             0
dtype: int64

In [18]:
# Выводим информацию по games
games.info()

<class 'pandas.core.frame.DataFrame'>
Index: 16954 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  object 
 1   platform         16954 non-null  object 
 2   year_of_release  16954 non-null  float64
 3   genre            16954 non-null  object 
 4   na_sales         16954 non-null  float64
 5   eu_sales         16954 non-null  object 
 6   jp_sales         16954 non-null  object 
 7   other_sales      16954 non-null  float64
 8   critic_score     16954 non-null  float64
 9   user_score       16954 non-null  object 
 10  rating           16954 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.6+ MB


Пропуски обработаны до корректировки типа данных

### Корректировка типа данных

Для оптимизации возможно применение числовых типов с уменьшенной разрядностью

In [19]:
# Преобразовываем типы данных с понижением разрядности
for col in ['na_sales',
               'eu_sales',
               'jp_sales',
               'other_sales',
               'critic_score',
               'user_score']:
    games[col] = pd.to_numeric(games[col], errors='coerce', downcast='float')

In [20]:
# Преобразовываем типы данных с понижением разрядности
for col in ['year_of_release']:
    games[col] = pd.to_numeric(games[col], errors='coerce', downcast='integer')

In [21]:
# Выводим информацию по games
games.info()

<class 'pandas.core.frame.DataFrame'>
Index: 16954 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  object 
 1   platform         16954 non-null  object 
 2   year_of_release  16954 non-null  int16  
 3   genre            16954 non-null  object 
 4   na_sales         16954 non-null  float32
 5   eu_sales         16948 non-null  float32
 6   jp_sales         16950 non-null  float32
 7   other_sales      16954 non-null  float32
 8   critic_score     16954 non-null  float32
 9   user_score       14490 non-null  float32
 10  rating           16954 non-null  object 
dtypes: float32(6), int16(1), object(4)
memory usage: 1.1+ MB


Тип данных в нужных столбцах изменены.

Обнаружились пропуски после смены типа данных на float32

In [22]:
# Пропуски в ['eu_sales', 'jp_sales', 'other_sales'] заполняем медианой по категориям платформы и году выпуска
for col in ['eu_sales', 'jp_sales', 'other_sales']:
    games[col] = games[col].fillna(
        games.groupby(['platform', 'year_of_release'])[col].transform('median')
    )

In [23]:
# Заполняем рейтинги значением '-1'
games['user_score'] = games['user_score'].fillna(-1)

In [24]:
# Выводим информацию о пропусках
games.isna().sum()

name               0
platform           0
year_of_release    0
genre              0
na_sales           0
eu_sales           0
jp_sales           0
other_sales        0
critic_score       0
user_score         0
rating             0
dtype: int64

In [25]:
# Выводим информацию о games
games.info()

<class 'pandas.core.frame.DataFrame'>
Index: 16954 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  object 
 1   platform         16954 non-null  object 
 2   year_of_release  16954 non-null  int16  
 3   genre            16954 non-null  object 
 4   na_sales         16954 non-null  float32
 5   eu_sales         16954 non-null  float32
 6   jp_sales         16954 non-null  float32
 7   other_sales      16954 non-null  float32
 8   critic_score     16954 non-null  float32
 9   user_score       16954 non-null  float32
 10  rating           16954 non-null  object 
dtypes: float32(6), int16(1), object(4)
memory usage: 1.1+ MB


Пропуски обработаны

### Явные и неявные дубликаты в данных

#### Исследуем данные на предмет неявных дубликатоы

In [26]:
# Уникальные значения в столбце 'genre'
games['genre'].unique()

array(['Sports', 'Platform', 'Racing', 'Role-Playing', 'Puzzle', 'Misc',
       'Shooter', 'Simulation', 'Action', 'Fighting', 'Adventure',
       'Strategy', 'MISC', 'ROLE-PLAYING', 'RACING', 'ACTION', 'SHOOTER',
       'FIGHTING', 'SPORTS', 'PLATFORM', 'ADVENTURE', 'SIMULATION',
       'PUZZLE', 'STRATEGY'], dtype=object)

In [27]:
# Приводим значения столбца 'genre' к нижнему регистру
games['genre'] = games['genre'].str.lower()

In [28]:
# Уникальные значения в столбце 'genre' после обработки. Неявные дубликаты убраны
games['genre'].unique()

array(['sports', 'platform', 'racing', 'role-playing', 'puzzle', 'misc',
       'shooter', 'simulation', 'action', 'fighting', 'adventure',
       'strategy'], dtype=object)

In [29]:
# Уникальные значения в столбце 'platform'.
games['platform'].unique()

array(['Wii', 'NES', 'GB', 'DS', 'X360', 'PS3', 'PS2', 'SNES', 'GBA',
       'PS4', '3DS', 'N64', 'PS', 'XB', 'PC', '2600', 'PSP', 'XOne',
       'WiiU', 'GC', 'GEN', 'DC', 'PSV', 'SAT', 'SCD', 'WS', 'NG', 'TG16',
       '3DO', 'GG', 'PCFX'], dtype=object)

In [30]:
# Уникальные значения в столбце 'rating'.
games['rating'].unique()

array(['E', '-1', 'M', 'T', 'E10+', 'K-A', 'AO', 'EC', 'RP'], dtype=object)

In [31]:
# Уникальные значения в столбце 'name'
games['name'].unique()

array(['Wii Sports', 'Super Mario Bros.', 'Mario Kart Wii', ...,
       'Woody Woodpecker in Crazy Castle 5', 'LMA Manager 2007',
       'Haitaka no Psychedelica'], shape=(11559,), dtype=object)

In [32]:
# Количество уникальных значений в столбце 'name'
games['name'].nunique()

11559

In [33]:
# Приводим значения столбца 'name' к нижнему регистру
games['name'] = games['name'].str.lower()

In [34]:
# Количество уникальных значений в столбце 'name'. Неявные дубликаты не обнаружены.
games['name'].nunique()

11559

#### Исследуем данные на предмет явных дубликатоы

In [35]:
# Сортируем датафрейм
games = games.sort_values(by=list(games.columns))

In [36]:
# Выводим количество дубликатов
games.duplicated().sum()

np.int64(241)

In [37]:
# Удаляем явные дубликаты
games = games.drop_duplicates()
games.info()

<class 'pandas.core.frame.DataFrame'>
Index: 16713 entries, 15191 to 9260
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16713 non-null  object 
 1   platform         16713 non-null  object 
 2   year_of_release  16713 non-null  int16  
 3   genre            16713 non-null  object 
 4   na_sales         16713 non-null  float32
 5   eu_sales         16713 non-null  float32
 6   jp_sales         16713 non-null  float32
 7   other_sales      16713 non-null  float32
 8   critic_score     16713 non-null  float32
 9   user_score       16713 non-null  float32
 10  rating           16713 non-null  object 
dtypes: float32(6), int16(1), object(4)
memory usage: 1.1+ MB


Было обнаружено 241 явных дубликатов. Данные строки были удалены.

### Промежуточный вывод по удалённым данным

В процессе предобработки датафрейма `games` были удалены строки по двум причинам:

| Этап | Действие | Удалено строк | Доля от исходного объёма |
|------|----------|---------------|--------------------------|
| 1 | Удаление строк с пропусками в `name` (и, соответственно, в `genre`) | 2 | 0.012% |
| 2 | Удаление явных дубликатов (`drop_duplicates()`) | 241 | 1.42% |
| **Итого** | | **243** | **≈ 1.43%** |

**Исходный размер:** 16 956 строк.
**Итоговый размер:** 16 713 строк.

**Относительные потери:**

```
243 / 16956 * 100 ≈ 1.43%
```

Потеря около 1.4% данных незначительна и не влияет на репрезентативность выборки. Основную долю удалённых строк составляют явные дубликаты (241 из 243), которые могли возникнуть при объединении источников данных и искажали бы статистику продаж и оценок. Единичные удаления строк без названия игры (2 строки) обусловлены невозможностью идентифицировать запись.

## Общий промежуточный вывод по предобработке

В ходе предобработки датасета `new_games.csv` были выполнены следующие шаги:

1. **Названия столбцов** приведены к «змеиному» регистру (`snake_case`) для удобства работы.
2. **Пропуски обработаны:**
   - строки без `name` (и `genre`) удалены (2 шт.);
   - `year_of_release` заполнен медианой по платформе;
   - `critic_score`, `user_score`, `rating` заполнены маркером `-1` как признаком отсутствия оценки;
   - пропуски в `eu_sales`, `jp_sales`, `other_sales`, появившиеся после смены типа, заполнены медианой по (`platform`, `year_of_release`).
3. **Типы данных оптимизированы:** числовые столбцы приведены к `float32`, `year_of_release` — к `int16`.
4. **Неявные дубликаты устранены:** значения `genre` приведены к нижнему регистру.
5. **Явные дубликаты удалены:** 241 строка.
6. **Итоговый объём:** 16 713 строк, 11 столбцов, пропусков нет.

Датафрейм готов к дальнейшему анализу: отбору периода 2000–2013, категоризации оценок, выделению топ-7 платформ и исследованию региональных предпочтений.

## Фильтрация данных

In [38]:
# Создаем новый датафрейм на основе среза games по годам (2000 - 2013)
df_actual = games[(games['year_of_release'] >= 2000) & (games['year_of_release'] <= 2013)].copy()
df_actual.head()

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating
3394,frozen: olaf's quest,3DS,2013,platform,0.27,0.27,0.00,0.05,-1.0,-1.0,-1
3906,frozen: olaf's quest,DS,2013,platform,0.21,0.26,0.00,0.04,-1.0,-1.0,-1
2478,tales of xillia 2,PS3,2012,role-playing,0.20,0.12,0.45,0.07,71.0,7.9,T
8460,.hack//g.u. vol.1//rebirth,PS2,2006,role-playing,0.00,0.00,0.17,0.00,-1.0,-1.0,-1
7182,.hack//g.u. vol.2//reminisce,PS2,2006,role-playing,0.11,0.09,0.00,0.03,-1.0,-1.0,-1


## Категоризация данных

In [39]:
# Категоризация данных по оценкам критиков
def categ_critic(value):
    if value >= 0 and value < 30:
        return 'низкая оценка'
    elif value >= 30 and value < 80:
        return 'средняя оценка'
    elif value >= 80 and value <= 100:
        return 'высокая оценка'
    else:
        return 'нет оценки'

df_actual['group_critic_score'] = df_actual['critic_score'].apply(categ_critic)

In [40]:
# Выводим количество игр по категориям оценки критиков
df_actual.groupby('group_critic_score')['name'].count()

group_critic_score
высокая оценка    1724
нет оценки        5703
низкая оценка       58
средняя оценка    5536
Name: name, dtype: int64

In [41]:
# Категоризация данных по оценкам игроков
def categ_user(value):
    if value >= 0 and value < 3:
        return 'низкая оценка'
    elif value >= 3 and value < 8:
        return 'средняя оценка'
    elif value >= 8 and value <= 10:
        return 'высокая оценка'
    else:
        return 'нет оценки'

df_actual['group_user_score'] = df_actual['user_score'].apply(categ_user)

In [42]:
# Выводим количество игр по категориям оценки игроков
df_actual.groupby('group_user_score')['name'].count()

group_user_score
высокая оценка    2328
нет оценки        6415
низкая оценка      119
средняя оценка    4159
Name: name, dtype: int64

In [43]:
# Выводим информацию по df_actual
df_actual.info()    

<class 'pandas.core.frame.DataFrame'>
Index: 13021 entries, 3394 to 9260
Data columns (total 13 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   name                13021 non-null  object 
 1   platform            13021 non-null  object 
 2   year_of_release     13021 non-null  int16  
 3   genre               13021 non-null  object 
 4   na_sales            13021 non-null  float32
 5   eu_sales            13021 non-null  float32
 6   jp_sales            13021 non-null  float32
 7   other_sales         13021 non-null  float32
 8   critic_score        13021 non-null  float32
 9   user_score          13021 non-null  float32
 10  rating              13021 non-null  object 
 11  group_critic_score  13021 non-null  object 
 12  group_user_score    13021 non-null  object 
dtypes: float32(6), int16(1), object(6)
memory usage: 1.0+ MB


In [44]:
# Первые 10 строк df_actual
df_actual.head(10)

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating,group_critic_score,group_user_score
3394,frozen: olaf's quest,3DS,2013,platform,0.27,0.27,0.00,0.05,-1.0,-1.0,-1,нет оценки,нет оценки
3906,frozen: olaf's quest,DS,2013,platform,0.21,0.26,0.00,0.04,-1.0,-1.0,-1,нет оценки,нет оценки
2478,tales of xillia 2,PS3,2012,role-playing,0.20,0.12,0.45,0.07,71.0,7.9,T,средняя оценка,средняя оценка
8460,.hack//g.u. vol.1//rebirth,PS2,2006,role-playing,0.00,0.00,0.17,0.00,-1.0,-1.0,-1,нет оценки,нет оценки
7182,.hack//g.u. vol.2//reminisce,PS2,2006,role-playing,0.11,0.09,0.00,0.03,-1.0,-1.0,-1,нет оценки,нет оценки
8719,.hack//g.u. vol.2//reminisce (jp sales),PS2,2006,role-playing,0.00,0.00,0.16,0.00,-1.0,-1.0,-1,нет оценки,нет оценки
8410,.hack//g.u. vol.3//redemption,PS2,2007,role-playing,0.00,0.00,0.17,0.00,-1.0,-1.0,-1,нет оценки,нет оценки
1575,.hack//infection part 1,PS2,2002,role-playing,0.49,0.38,0.26,0.13,75.0,8.5,T,средняя оценка,высокая оценка
9189,.hack//link,PSP,2010,role-playing,0.00,0.00,0.14,0.00,-1.0,-1.0,-1,нет оценки,нет оценки
3023,.hack//mutation part 2,PS2,2002,role-playing,0.23,0.18,0.20,0.06,76.0,8.9,T,средняя оценка,высокая оценка


In [45]:
# ТОП-7 платформ по количеству игр
df_actual['platform'].value_counts().sort_values(ascending=False).head(7)

platform
PS2     2161
DS      2150
Wii     1309
PSP     1196
X360    1151
PS3     1112
XB       824
Name: count, dtype: int64

## Основной вывод и описание проделанной работы

### Что было сделано в проекте

В рамках проекта команды «Секреты Темнолесья» проведена полная предобработка и подготовка датасета `new_games.csv` для дальнейшего исследования игровой индустрии 2000–2013 годов. Работа включала следующие этапы:

#### 1. Знакомство с данными
- Загружен датасет из 11 столбцов и **16 956 строк**.
- Изучены типы данных: часть числовых столбцов (`eu_sales`, `jp_sales`, `user_score`) хранилась как `object`, что требовало приведения к числовому типу.
- Выявлены пропуски в `name` (2), `year_of_release` (275), `critic_score` (8714), `user_score` (6804), `rating` (6871).

#### 2. Предобработка данных
- **Метки столбцов** приведены к `snake_case` (`Year of Release` → `year_of_release` и т. д.).
- **Пропуски:**
  - строки без `name` (и `genre`) удалены (2 шт.);
  - `year_of_release` заполнен **медианой** по платформе;
  - `critic_score`, `user_score`, `rating` заполнены маркером `-1` как признак отсутствия оценки;
  - пропуски в `eu_sales`, `jp_sales`, `other_sales`, возникшие после смены типа, заполнены **медианой** по (`platform`, `year_of_release`).
- **Типы данных оптимизированы:** `na_sales`, `eu_sales`, `jp_sales`, `other_sales`, `critic_score`, `user_score` → `float32`; `year_of_release` → числовой тип (после отбора среза хранится как `float64`, но по смыслу — целочисленный).
- **Неявные дубликаты:** значения `genre` приведены к нижнему регистру.
- **Явные дубликаты:** удалено 241 полных дубликатов строк.
- **Итог предобработки:** 16 713 строк без пропусков.

**Удалено всего:** 243 строки (2 из-за пропусков в `name` + 241 дубликат), что составляет **≈1.43%** от исходного объёма — потеря незначительна и не влияет на репрезентативность выборки.

#### 3. Формирование среза данных
Создан новый датафрейм `df_actual`, содержащий только игры, выпущенные в период **с 2000 по 2013 год включительно**:

```python
df_actual = games[(games['year_of_release'] >= 2000) & (games['year_of_release'] <= 2013)].copy()
```

**Итоговый размер среза:** **13 021 строк** (без пропусков), 13 столбцов.

### Новые поля, добавленные в датасет

В `df_actual` добавлены **два категориальных признака**, полученных на основе числовых оценок:

| Поле | Источник | Правила категоризации |
|------|----------|------------------------|
| `group_critic_score` | `critic_score` (0–100) | `низкая оценка` — [0; 30); `средняя оценка` — [30; 80); `высокая оценка` — [80; 100]; `нет оценки` — для значения `-1` |
| `group_user_score` | `user_score` (0–10) | `низкая оценка` — [0; 3); `средняя оценка` — [3; 8); `высокая оценка` — [8; 10]; `нет оценки` — для значения `-1` |

Эти поля позволяют агрегировать продажи и строить сравнительный анализ по группам оценок критиков и игроков, а также отдельно исследовать игры без оценок (маркер `-1`).

#### Распределение игр по категориям оценок критиков

| Категория `group_critic_score` | Количество игр |
|--------------------------------|---------------:|
| высокая оценка                 | 1724           |
| средняя оценка                 | 5536           |
| низкая оценка                  | 58             |
| нет оценки                     | 5703           |
| **Итого**                      | **13021**      |

#### Распределение игр по категориям оценок пользователей

| Категория `group_user_score` | Количество игр |
|------------------------------|---------------:|
| высокая оценка               | 2328           |
| средняя оценка               | 4159           |
| низкая оценка                | 119            |
| нет оценки                   | 6415           |
| **Итого**                    | **13021**      |


### Топ-7 платформ за 2000–2013 годы

По количеству выпущенных игр в срезе `df_actual` лидируют:

| Платформа | Кол-во игр |
|-----------|-----------:|
| PS2 | 2161 |
| DS | 2150 |
| Wii | 1309 |
| PSP | 1196 |
| X360 | 1151 |
| PS3 | 1112 |
| XB | 824 |

Это платформы-лидеры эпохи, и именно по ним имеет смысл проводить основной анализ в финальной статье: динамику продаж по годам, региональные предпочтения и жанровые тренды.

### Итоговый вывод

Датасет полностью подготовлен к аналитическому этапу:

- **Объём среза:** 13 021 игр периода 2000–2013 гг., пропусков нет.
- **Оптимизированные типы данных** и приведённые метки столбцов.
- **Удалены явные дубликаты и неявные расхождения** в категориальных значениях (`genre`, `name`).
- **Добавлены новые признаки** `group_critic_score` и `group_user_score` для категоризации игр по оценкам.
- **Определены 7 ключевых платформ** эпохи, на которых будет строиться основное исследование.

Такой срез позволяет корректно решать задачи проекта: анализировать эволюцию жанров, сравнивать платформы по продажам, изучать региональные предпочтения (NA/EU/JP) и оценивать, как пользовательские и экспертные оценки связаны с коммерческим успехом игр, в том числе в любимом жанре команды — RPG.